### PR TITLE
Escape HTML special characters in HTML sidebar entries (2nd take)

### LINKED ISSUE #1244: Suggestions wrongly deal with escaped chars
If you type in the search box "&" you will get the suggestion of functions that their name has been escaped such as "&lt;/2". Same if you type "gt".

I think we should keep the original name and the escaped name.

### PR DESCRIPTION
**Pull request by eksperimental**
This is a second take to fix what was intented by d26ca71417c40f3e5d9ab49d0933b277c983962d
Superseeds #1227
Closes #1244

### CODE REVIEW

**Comment by eksperimental** [comment]
Needs reviewing from JS programmers
/cc @davydog187 @archdragon

File: assets/js/autocomplete/display.js [hunk 1]
```diff
@@ -16,7 +17,7 @@ function updateSuggestions (term) {
   const template = autocompleteResultsTemplate({
     empty: results.length === 0,
     results: results,
-    term: term
+    term: escapeHtmlEntities(term)
```
Comment (archdragon) [inline assets/js/autocomplete/display.js:20]: With the current escaping I would remove this escape and use 

```
<div class="title">"<em>{{term}}</em>"</div>
```

in `autocomplete-suggestions.js` to escape the term

File: assets/js/autocomplete/suggestions.js [hunk 2]
```diff
@@ -151,7 +151,13 @@ function findIn (elements, term, categoryName) {
...
  */
 function highlight (match) {
-  return match.input.replace(match, `<em>${match[0]}</em>`)
+  const EmOpen = '###EM_OPEN###'
+  const EmClose = '###EM_CLOSE###'
+
+  let result = match.input
+  result = result.replace(match, EmOpen + match[0] + EmClose)
+  result = escapeHtmlEntities(result)
+  return result.replace(EmOpen, '<em>').replace(EmClose, '<\/em>')
```
Comment (archdragon) [inline assets/js/autocomplete/suggestions.js:160]: I think this is not needed - I'm not getting any escpaing issues with the default implementation.

File: assets/js/templates/sidebar-items.handlebars [hunk 3]
```diff
@@ -62,7 +62,7 @@
               <ul class="{{group.key}}-list deflist">
                 {{#each group.nodes}}
                   <li>
-                    <a href="{{node.id}}.html#{{anchor}}">{{{id}}}</a>
+                    <a href="{{node.id}}.html#{{anchor}}">{{id}}</a>
```
Comment (archdragon) [inline assets/js/templates/sidebar-items.handlebars:65]: 👍

**Comment by eksperimental** [comment]
> With the current change on the backend, I think it's enough to just replace:
> 
> `{{{id}}}` with `{{id}}`
> 
> I think with the current backend changes this will take care of ID escaping and it will be enough to fix issue described.
> 
> I've tested just escaping `{{id}}` on my end and all works well then.
> 
> Does that make sense? Or have I missed something?

If you use double curly brackets the then it will be escaped twice.

**Comment by eksperimental** [comment]
> With the current change on the backend, I think it's enough to just replace:
by backend you mean the Elixir backend or JavaScript?

File: assets/js/autocomplete/suggestions.js (hunk 2 above)
Comment (eksperimental) [reply assets/js/autocomplete/suggestions.js:160]: The tricky part is when the search term contains a special html entity, such as `<` or `>`. Could you please try them with that. The changes in this functions is to deal with that edge case.

**Comment by archdragon** [comment]
@eksperimental Sorry, I was not being precise!

I meat elixir changes in the current PR + just escaping id in `JavaScript`.

So to sum up I think that changing
```
 nodes = Enum.map(docs, fn doc -> %{id: h(doc.id), anchor: URI.encode(HTML.link_id(doc))} end)
```
to
```
nodes = Enum.map(docs, fn doc -> %{id: doc.id, anchor: URI.encode(HTML.link_id(doc))} end)
```

and
```
{{{id}}}
```

to

```
{{id}}
```

should be enough. WDYT?

File: assets/js/autocomplete/suggestions.js (hunk 2 above)
Comment (archdragon) [reply assets/js/autocomplete/suggestions.js:160]: I see - I can break it with `">` thought we already account for that, but it's not the case.

image

thanks!

**Comment by archdragon** [comment]
After reading the discussion comments, I see where extra escaping is needed. Adding my +1 :)

**Comment by eksperimental** [comment]
> After reading the discussion comments, I see where extra escaping is needed. Adding my +1 :)

Thank you. Still I am considering one of your suggestions to simplify the code.

File: assets/js/autocomplete/display.js (hunk 1 above)
Comment (eksperimental) [reply assets/js/autocomplete/display.js:20]: This suggestion can be implemented.

Comment (eksperimental) [reply assets/js/autocomplete/display.js]: done

File: assets/js/autocomplete/suggestions.js (hunk 2 above)
Comment (eksperimental) [inline assets/js/autocomplete/suggestions.js:160]: @archdragon do you know of a more idiomatic way of doing this in JS?

**Comment by josevalim** [comment]
Absolutely beautiful, thank you! :green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by eksperimental** [comment]
> Absolutely beautiful, thank you! green_heart blue_heart purple_heart yellow_heart heart

You are welcome!

**Pull request merged**